### PR TITLE
Theme adds support for dicts

### LINKED ISSUE #400: It would be nice to pass in the theme as a variable defined rather than a file
**Is your feature request related to a problem? Please describe.**
So that users cannot change the themes in my game once packaged, I'd like to be able to pass in a theme as a variable defined in Python

**Describe the solution you'd like**
Example in my games_themes.py:
```json

game_themes = {
    "defaults": {
            "colours": {
                    "normal_bg": "#45494e",
                    "hovered_bg": "#35393e",
                    "disabled_bg": "#25292e",
                    "selected_bg": "#193754",
                    "dark_bg": "#15191e",
                    "normal_text": "#c5cbd8",
                    "hovered_text": "#FFFFFF",
                    "selected_text": "#FFFFFF",
                    "disabled_text": "#6d736f",
                    "link_text": "#0000EE",
                    "link_hover": "#2020FF",
                    "link_selected": "#551A8B",
                    "text_shadow": "#777777",
                    "normal_border": "#DDDDDD",
                    "hovered_border": "#B0B0B0",
                    "disabled_border": "#808080",
                    "selected_border": "#8080B0",
                    "active_border": "#8080B0",
                    "filled_bar": "#f4251b",
                    "unfilled_bar": "#CCCCCC"
                }
        },
    "button": {
        "misc": {
                "border_width": "1",
                "shadow_width": "2"
        }
    }
}
```

### PR DESCRIPTION
**Pull request by LondonClass**
fixes #400

I found that update_theming and load_theme perform similar functions... Do they need to be refactored?

My other goal is to enable UIManager to dynamically replace ui_theme during runtime. If this is necessary, I may try to do it in the future.

### CODE REVIEW

**Comment by MyreMylar** [comment]
Cool, looks like this could just use a quick unit test of changing a theme by passing in a dictionary instead of a theme file.

> My other goal is to enable UIManager to dynamically replace ui_theme during runtime. If this is necessary, I may try to do it in the future.

You should be able to do this fairly easily in the UIMananger with something like:

```python
def load_and_update_theming(theme_file_path: str)
  # load theme data
  self.ui_theme.load_theme(theme_file_path)
  # update UIElements
  for sprite in self.ui_group.sprites():
      sprite.rebuild_from_changed_theme_data()
```

Or at least it used to work like that at one point because you could edit the theme files while the apps were running and see the changes live. Maybe it still does!

**Comment by LondonClass** [comment]
@MyreMylar updata_theming and load_theme seem to be doing the same thing. I want to remove the update_theming method and merge its functionality into the load_theme.

**Comment by LondonClass** [comment]
> You should be able to do this fairly easily in the UIMananger with something like:
> 
> ```python
> def load_and_update_theming(theme_file_path: str)
>   # load theme data
>   self.ui_theme.load_theme(theme_file_path)
>   # update UIElements
>   for sprite in self.ui_group.sprites():
>       sprite.rebuild_from_changed_theme_data()
> ```
> 
> Or at least it used to work like that at one point because you could edit the theme files while the apps were running and see the changes live. Maybe it still does!

My idea is to save multiple UIAppearanceTheme instances and dynamically replace them at runtime, so there is no need to read files repeatedly.

File: pygame_gui/core/interfaces/appearance_theme_interface.py [hunk 1]
```diff
@@ -147,12 +147,12 @@ def get_colour_or_gradient(self, colour_id: str,
         """
 
     @abstractmethod
-    def load_theme(self, file_path: Union[str, PathLike, io.StringIO]):
+    def load_theme(self, file_path: Union[str, os.PathLike, io.StringIO, PackageResource, dict]):
```
Comment (MyreMylar) [inline pygame_gui/core/interfaces/appearance_theme_interface.py:150]:
```suggestion
    def load_theme(self, file_path: Union[str, PathLike, io.StringIO, PackageResource, dict]):
```

File: pygame_gui/core/interfaces/appearance_theme_interface.py [hunk 2]
```diff
@@ -147,12 +148,12 @@ def get_colour_or_gradient(self, colour_id: str,
         """
 
     @abstractmethod
-    def load_theme(self, file_path: Union[str, PathLike, io.StringIO]):
+    def load_theme(self, file_path: Union[str, PathLike, io.StringIO, PackageResource, dict]):
```
Comment (MyreMylar) [inline pygame_gui/core/interfaces/appearance_theme_interface.py:151]:
```suggestion
    def load_theme(self, file_path: Union[str, PathLike, io.StringIO, 'PackageResource', dict]):
```

File: pygame_gui/core/interfaces/appearance_theme_interface.py [hunk 3]
```diff
@@ -10,6 +10,7 @@
 from pygame_gui.core.interfaces.font_dictionary_interface import IUIFontDictionaryInterface
 from pygame_gui.core.interfaces.colour_gradient_interface import IColourGradientInterface
 from pygame_gui.core.interfaces.gui_font_interface import IGUIFontInterface
+from pygame_gui.core.utility import PackageResource
```
Comment (MyreMylar) [inline pygame_gui/core/interfaces/appearance_theme_interface.py:13]:
```suggestion
from pygame_gui.core.utility import PackageResource
```

File: tests/test_core/test_ui_appearance_theme.py [hunk 4]
```diff
@@ -18,6 +18,13 @@ def test_load_non_default_theme_from_package(self, _init_pygame, _display_surfac
         colour = theme.get_colour(colour_id='dark_bg', combined_element_ids=['text_box'])
         assert colour == pygame.Color('#25f92e')
 
+    def test_load_non_default_theme_from_dictionary(self, _init_pygame, _display_surface_return_none):
+        theme = UIAppearanceTheme(BlockingThreadedResourceLoader(), locale='en')
+        theme_dict = {"text_box": {"colours": {"dark_bg": "#25f92e"}}}
+        theme.load_theme()
```
Comment (MyreMylar) [inline tests/test_core/test_ui_appearance_theme.py:24]:
```suggestion
        theme.load_theme(theme_dict)
```

**Comment by MyreMylar** [comment]
just need to add a test for update theming and this should pass.

**Comment by LondonClass** [comment]
> My idea is to save multiple UIAppearanceTheme instances and dynamically replace them at runtime, so there is no need to read files repeatedly.

Done this.

**Review by MyreMylar — APPROVED** [review]
OK, LGTM 👍

**Pull request merged**